### PR TITLE
Support sdks v1.0

### PR DESCRIPTION
**Pull request by svlaev**
https://github.com/ccm-innovation/react-native-twilio-chat/pull/24

### CODE REVIEW

**Comment by svlaev** [comment]
@bradbumbalough here you go :)

**Comment by benoist** [comment]
I tried to install this, but it's requiring React via pods and it cant install the latest version of React.
In my previous PR I removed the pod references completely, any reason this is still a requirement?

**Comment by benoist** [comment]
I see in the example the react extension is also included from the pod file, but thats not required as react-native link will link include the project. For this to work, the search paths have to be fixed.

**Comment by svlaev** [comment]
@benoist Just merged your changes into the PR. 
IMHO everything should be fine now

**Comment by benoist** [comment]
Ok I will try your changes 👍

File: lib/index.js [hunk 1]
```diff
@@ -1,7 +1,8 @@
 exports.AccessManager = require('./AccessManager').default;
 exports.Client = require('./Client').default;
 exports.Channel = require('./Channel').default;
-exports.UserInfo = require('./UserInfo').default;
+exports.User = require('./User').default;
```
Comment (benoist) [inline lib/index.js]: Some files still have references to userinfo

File: lib/Member.js [hunk 2]
```diff
@@ -1,11 +1,22 @@
-import UserInfo from './UserInfo';
+const {
```
Comment (benoist) [inline lib/Member.js]: It's missing an import  NativeModules from react native

**Review by benoist — COMMENTED** [review]
In Client.js on iOS it's complaining the following constant does not exist:
Constants.TCHClientSynchronizationStrategy.ChannelsList

File: ios/RCTTwilioChat/RCTConvert+TwilioChatClient.m [hunk 3]
```diff
@@ -38,17 +38,12 @@ @implementation RCTConvert (TwilioChatClient)
...
                                      }), TCHChannelTypePublic, integerValue)
 
-RCT_ENUM_CONVERTER(TCHUserInfoUpdate,(@{
-                                        @"FriendlyName" : @(TCHUserInfoUpdateFriendlyName),
-                                        @"Attributes" : @(TCHUserInfoUpdateAttributes),
-                                        @"ReachabilityOnline": @(TCHUserInfoUpdateReachabilityOnline),
-                                        @"ReachabilityNotifiable": @(TCHUserInfoUpdateReachabilityNotifiable),
-                                        }), TCHUserInfoUpdateFriendlyName, integerValue)
-
-RCT_ENUM_CONVERTER(TCHClientSynchronizationStrategy,(@{
```
Comment (benoist) [inline ios/RCTTwilioChat/RCTConvert+TwilioChatClient.m:48]: This is removed, but Client.js still depends on it.

Comment (svlaev) [reply ios/RCTTwilioChat/RCTConvert+TwilioChatClient.m:48]: Fixed: https://github.com/ccm-innovation/react-native-twilio-chat/pull/25/commits/9d048da5d8083a65c1313ab490d1949ab50ea0c8

File: lib/Member.js (hunk 2 above)
Comment (svlaev) [reply lib/Member.js]: Fixed in: https://github.com/ccm-innovation/react-native-twilio-chat/pull/25/commits/9a62714ee9f8c2a48be22f03ab7ad912adf1012a

File: lib/index.js (hunk 1 above)
Comment (svlaev) [reply lib/index.js]: Fixed in: https://github.com/ccm-innovation/react-native-twilio-chat/pull/25/commits/9a62714ee9f8c2a48be22f03ab7ad912adf1012a

**Comment by svlaev** [comment]
@benoist thanks for the feedback! :)

**Comment by benoist** [comment]
Hmm for some reason synchronizationStatusChanged is not fired anymore...

>@benoist thanks for the feedback! :)

No problem, you did a great job already!!

File: ios/RCTTwilioChat/RCTTwilioChatClient.m [hunk 4]
```diff
@@ -158,7 +158,7 @@ - (void)chatClient:(TwilioChatClient *)client connectionStateChanged:(TCHClientC
                                                  body:@(state)];
 }
 
-- (void)chatClient:(TwilioChatClient *)client synchronizationStatusChanged:(TCHClientSynchronizationStatus)status {
+- (void)chatClient:(TwilioChatClient *)client synchronizationStatusUpdated:(TCHClientSynchronizationStatus)status {
```
Comment (svlaev) [inline ios/RCTTwilioChat/RCTTwilioChatClient.m]: @benoist this is the reason why the `synchronizationStatusChanged ` wasn't called.
Twilio changed the delegate call

**Comment by bradbyte** [comment]
@benoist @svlaev thanks so much for working on this! I'm sorry I was out of pocket.

**Comment by benoist** [comment]
No worries, I will test in again in a couple of hours.

@svlaev is this change for iOS only? I see an override in android as well with the old name
https://github.com/svlaev/react-native-twilio-chat/blob/dafe3e9d0f74e7d514d570366da99d1eaa48eadb/android/src/main/java/com/bradbumbalough/RCTTwilioChat/RCTTwilioChatClient.java#L333

**Comment by bradbyte** [comment]
Looks like the Android 1.0 still has `onSynchronizationChanged` as the method name, but we should update the event name so it syncs with the iOS change. https://media.twiliocdn.com/sdk/android/chat/releases/1.0.0/docs/com/twilio/chat/ChatClient.html#onSynchronizationChanged-com.twilio.chat.Channel-

**Comment by svlaev** [comment]
@benoist 
> @svlaev is this change for iOS only? I see an override in android as well with the old name
> https://github.com/svlaev/react-native-twilio-chat/blob/dafe3e9d0f74e7d514d570366da99d1eaa48eadb/android/src/main/java/com/bradbumbalough/RCTTwilioChat/RCTTwilioChatClient.java#L333

I believe there's a difference between client sync status change and channel sync status change.
I found this method in the adnroid sdk: https://media.twiliocdn.com/sdk/android/chat/releases/1.0.0/docs/com/twilio/chat/ChatClientListener.html#onClientSynchronization-com.twilio.chat.ChatClient.SynchronizationStatus-
which to me corresponds to https://media.twiliocdn.com/sdk/ios/chat/releases/1.0.0/docs/Protocols/TwilioChatClientDelegate.html#//api/name/chatClient:synchronizationStatusUpdated:

So to me both SDKs will emit the same event to the RN lib

@bradbumbalough I'll change the name in a jiffy

**Comment by svlaev** [comment]
@bradbumbalough here it is https://github.com/ccm-innovation/react-native-twilio-chat/pull/25/commits/4f82da24c5a48e193f8347e0e84f8563bfaeb8e6 :)

**Comment by benoist** [comment]
There are some references left to userInfo in Client.js, but should be changed to user

In channels js
```js
initialize() {
    return TwilioChatChannels.synchronize(this.sid);
  }
```

but the synchronize is not there anymore. Should be updated to the correct one or the initialize method should be removed. I got it to work by removing it, but not sure if thats correct or not.

**Comment by benoist** [comment]
On android:

TwilioChatMembers.userDescriptor does not exist which is called from member.js

**Comment by svlaev** [comment]
Thanks again for the feedback! I will really appreciate it if somebody spends some time helping me with this.
I have other things to do now

**Comment by benoist** [comment]
I've created this https://github.com/svlaev/react-native-twilio-chat/pull/1
This contains a working example application working on android and iOS.
Just so it's easier to fix bugs and make it release ready it might be an idea to create a version branch for the current release version to perform patches on and then just merge this into master.
Once its ready and tested it can be released to npm

**Comment by svlaev** [comment]
I've merged https://github.com/svlaev/react-native-twilio-chat/pull/1 and it's included in this PR.
Thanks for helping me, @benoist 🍻

**Comment by benoist** [comment]
@bradbumbalough anything you want to see changed either in the current PR or the process of continuing development of the 1.0 SDK like I suggested?

**Comment by ukosodev** [comment]
Hi, Can I know when this pull request will be merged ? As I am using this library and the pull requests fix some issues I have currently. Thanks.

**Comment by Kabangi** [comment]
@svlaev @benoist How do you instantiate a channel after getting the channels with `getUserChannels()`. I have tried using `new Channel()` but I can't seem to figure out how to get members from this channel with getMembers() but this returns empty paginator all the time. Am using the support-v1-sdk branch from @svlaev fork.

**Comment by Kabangi** [comment]
Figured the `getMembers` on a channel has an issue. I have changed it to 
```
   async getMembers() {
    // There is a difference between this method and the iOS SDK method getMembers.
    // The iOS method is async and uses MemberPaginator
    // iOS SDK: - (void)membersWithCompletion:(TCHMemberPaginatorCompletion)completion
    // https://media.twiliocdn.com/sdk/ios/chat/releases/1.0.0/docs/Classes/TCHMembers.html#//api/name/membersWithCompletion:
    // Android SDK: public java.util.List<Member> getMembersList()
    // https://media.twiliocdn.com/sdk/android/chat/releases/1.0.0/docs/com/twilio/chat/Members.html#getMembersList--
    // Carl Olivier from Twilio team said this change will be addressed soon, so
    // this code is temporary
    if(Platform.OS === 'ios') {
      return TwilioChatMembers.getMembers(this.sid)
      .then(({ sid, type, paginator }) => new Paginator(sid, type, paginator));
    }
    const members = await TwilioChatMembers.getMembers(this.sid);
    var paginatorObj = {
      hasNextPage: false,
      items: members.items || []
    };
    // manual creation of the Paginator
    const paginator = new Paginator(members.sid, members.type, paginatorObj);
    return Promise.resolve(paginator);
  }
```
This returns an instance of `Paginator instance with Member items` whose methods  includes `getUserDescriptor()` which is not implemented atleast on android side.

**Comment by svlaev** [comment]
@Kabangi If by "How do you instantiate a channel after getting the channels with getUserChannels()" you mean you want to create a new channel, this is the method: https://github.com/ccm-innovation/react-native-twilio-chat/blob/1.0-rc1/lib/Client.js#L265

**Comment by Kabangi** [comment]
No I want to get an instance that I can append event listeners to. With the new sdk getUserChannels returns instance of Paginator with channel ChannelDescriptors unlike before. Meaning you have to either get an instance of the channel either by calling `client.getChannel(sid)` or using `channelDescriptor.getChannel()`. I resulted to using option one.

**Comment by svlaev** [comment]
@Kabangi `getUserChannels()` returns paginator. It has `items` property: https://github.com/ccm-innovation/react-native-twilio-chat/blob/1.0-rc1/lib/Paginator.js#L22
`items` is a simple array of channels. You can iterate it.

**Comment by Kabangi** [comment]
@svlaev Am using the support-v1-sdk branch from your @svlaev fork which has support for v1.0.0, that's why am discussing this on this pull request. Check my updated comment above.

**Comment by svlaev** [comment]
@Kabangi both approaches are fine. 
Please feel free to contribute to this PR if you think we can improve something :)

**Comment by Kabangi** [comment]
Sure I would love to but right now am not sure where to direct it to. Is it on your repo or should I wait for the merge to happen?

**Comment by svlaev** [comment]
@Kabangi TBH I don't know when the merge will happen, so you can PR to my PR just like @benoist did: https://github.com/ccm-innovation/react-native-twilio-chat/pull/25#issuecomment-301729520

**Comment by Kabangi** [comment]
@svlaev No sure if I did it right. merged your fork to mine then made some improvements and submitted a PR on the origin. #31

**Comment by Kabangi** [comment]
Also on this PR. There exists methods that return native Long to promises causing crashes when called. exampe include `getMessagesCount()`. 

```
@Override
   public void onSuccess(Long count) {
      promise.resolve(count);
    }
```
my solution right now is to convert the long to string then change it back to int on javascript side.
```
@Override
 public void onSuccess(Long count) {
       promise.resolve(Long.toString(count));
  }
```

I will send a PR with this if you guys think this is a good way to go around it.

File: lib/ChannelDescriptor.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    this.uniqueName = props.uniqueName;
+    this.attributes = props.attributes;
+    this.dateCreated = new Date(props.dateCreated);
+    this.dateUpdated = new Date(props.dateUpdated);
+    this.createdBy = props.createdBy;
+    if (props.membersCount) this.membersCount = props.membersCount;
+    if (props.messagesCount) this.messageCount = props.messagesCount;
+  }
+
+  getChannel() {
```
Comment (mattshen) [inline lib/ChannelDescriptor.js]: recommend return a Promise for information retrieval

File: lib/ChannelDescriptor.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    return TwilioChatChannels.getChannel(this.sid);
```
Comment (mattshen) [inline lib/ChannelDescriptor.js]: is returning a Promise better?

File: lib/Member.js [hunk 7]
```diff
@@ -1,11 +1,28 @@
...
-    this.userInfo = new UserInfo(props.userInfo);
+  constructor(props, channelSid) {
+    this.identity = props.identity;
+    this.channelSid = channelSid;
     this.lastConsumedMessageIndex = props.lastConsumedMessageIndex;
     this.lastConsumptionTimestamp = this.lastConsumptionTimestamp ? new Date(this.lastConsumptionTimestamp) : null;
   }
+
+  getUserDescriptor() {
+    return TwilioChatMembers.userDescriptor(this.channelSid, this.identity);
```
Comment (mattshen) [inline lib/Member.js]: returning a Promise?

File: lib/Member.js [hunk 8]
```diff
@@ -1,11 +1,28 @@
...
+  }
+
+  getAndSubscribeUser() {
+    return TwilioChatMembers.subscribedUser(this.channelSid, this.identity);
```
Comment (mattshen) [inline lib/Member.js]: returning a Promise?

**Comment by simonasdev** [comment]
I've added an issue #37  regarding the crash in `getMessagesCount` in this PR. The string workaround works, since we only need to construct an argument which React Native bridge is defined to parse (and `Long` isn't one of them), and use it as we will on Javascript side.

**Pull request closed without merging**